Error when merging whatsapp backups when there are corrupted databases (master)

When doing some test I found an error when processing the msgstore file inside a big UFDR. The strange thing is that the error does not occur when processing the msgstore.db files alone.
The following exception occurs

> org.apache.tika.exception.TikaException: WAExtractorException Exception
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parseDB(WhatsAppParser.java:340)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parseAndCheckIfIsMainDb(WhatsAppParser.java:351)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parse(WhatsAppParser.java:196)
	at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:280)
	at dpf.sp.gpinf.indexer.parsers.IndexerDefaultParser.parse(IndexerDefaultParser.java:243)
	at dpf.sp.gpinf.indexer.io.ParsingReader$BackgroundParsing.run(ParsingReader.java:247)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: dpf.mg.udi.gpinf.whatsappextractor.WAExtractorException: org.sqlite.SQLiteException: [SQLITE_NOTADB]  File opened that is not a database file (file is not a database)
	at dpf.mg.udi.gpinf.whatsappextractor.ExtractorAndroid.extractChatList(ExtractorAndroid.java:96)
	at dpf.mg.udi.gpinf.whatsappextractor.Extractor.getChatList(Extractor.java:28)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.extractChatList(WhatsAppParser.java:375)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parseDB(WhatsAppParser.java:335)
	... 10 more
Caused by: org.sqlite.SQLiteException: [SQLITE_NOTADB]  File opened that is not a database file (file is not a database)
	at org.sqlite.core.DB.newSQLException(DB.java:1012)
	at org.sqlite.core.DB.newSQLException(DB.java:1024)
	at org.sqlite.core.DB.throwex(DB.java:989)
	at org.sqlite.core.NativeDB.prepare_utf8(Native Method)
	at org.sqlite.core.NativeDB.prepare(NativeDB.java:134)
	at org.sqlite.core.DB.prepare(DB.java:257)
	at org.sqlite.jdbc3.JDBC3Statement.executeQuery(JDBC3Statement.java:66)
	at org.sqlite.jdbc3.JDBC3Statement.executeQuery(JDBC3Statement.java:56)
	at org.sqlite.jdbc3.JDBC3DatabaseMetaData.getTables(JDBC3DatabaseMetaData.java:1844)
	at org.sqlite.jdbc3.JDBC3DatabaseMetaData.getColumns(JDBC3DatabaseMetaData.java:1205)
	at dpf.mg.udi.gpinf.whatsappextractor.ExtractorAndroid.databaseHasSortTimestamp(ExtractorAndroid.java:105)
	at dpf.mg.udi.gpinf.whatsappextractor.ExtractorAndroid.extractChatList(ExtractorAndroid.java:67)

There is also an error in the msgstore file that I processed alone and it worked.

> org.apache.tika.exception.TikaException: WAExtractorException Exception
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parseDB(WhatsAppParser.java:340)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parseAndCheckIfIsMainDb(WhatsAppParser.java:351)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parse(WhatsAppParser.java:196)
	at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:280)
	at dpf.sp.gpinf.indexer.parsers.IndexerDefaultParser.parse(IndexerDefaultParser.java:243)
	at dpf.sp.gpinf.indexer.io.ParsingReader$BackgroundParsing.run(ParsingReader.java:247)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: dpf.mg.udi.gpinf.whatsappextractor.WAExtractorException: org.sqlite.SQLiteException: [SQLITE_CORRUPT]  The database disk image is malformed (database disk image is malformed)
	at dpf.mg.udi.gpinf.whatsappextractor.ExtractorAndroid.extractChatList(ExtractorAndroid.java:96)
	at dpf.mg.udi.gpinf.whatsappextractor.Extractor.getChatList(Extractor.java:28)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.extractChatList(WhatsAppParser.java:375)
	at dpf.mg.udi.gpinf.whatsappextractor.WhatsAppParser.parseDB(WhatsAppParser.java:335)
	... 10 more
Caused by: org.sqlite.SQLiteException: [SQLITE_CORRUPT]  The database disk image is malformed (database disk image is malformed)
	at org.sqlite.core.DB.newSQLException(DB.java:1012)
	at org.sqlite.core.DB.newSQLException(DB.java:1024)
	at org.sqlite.core.DB.throwex(DB.java:989)
	at org.sqlite.core.NativeDB.prepare_utf8(Native Method)
	at org.sqlite.core.NativeDB.prepare(NativeDB.java:134)
	at org.sqlite.core.DB.prepare(DB.java:257)
	at org.sqlite.jdbc3.JDBC3Statement.executeQuery(JDBC3Statement.java:66)
	at org.sqlite.jdbc3.JDBC3Statement.executeQuery(JDBC3Statement.java:56)
	at org.sqlite.jdbc3.JDBC3DatabaseMetaData.getTables(JDBC3DatabaseMetaData.java:1844)
	at org.sqlite.jdbc3.JDBC3DatabaseMetaData.getColumns(JDBC3DatabaseMetaData.java:1205)
	at dpf.mg.udi.gpinf.whatsappextractor.ExtractorAndroid.databaseHasSortTimestamp(ExtractorAndroid.java:105)
	at dpf.mg.udi.gpinf.whatsappextractor.ExtractorAndroid.extractChatList(ExtractorAndroid.java:67)
	... 13 more
2022-03-13 16:41:28	[WARN]	[indexer.parsers.IndexerDefaultParser]			ParsingThread-210 Parsing exception: TrisInIdem_Of0896-2020_Item_01.ufdr/Samsung GSM_SM-N975F Galaxy Note 10+.zip/data/data/com.whatsapp/databases/msgstore.db (2147483647 bytes)		dpf.mg.udi.gpinf.whatsappextractor.WAExtractorException: org.sqlite.SQLiteException: [SQLITE_CORRUPT]  The database disk image is malformed (database disk image is malformed)

--- a/iped-parsers/iped-parsers-impl/src/main/java/dpf/mg/udi/gpinf/whatsappextractor/ChatMerge.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/dpf/mg/udi/gpinf/whatsappextractor/ChatMerge.java
@@ -48,6 +48,9 @@ public class ChatMerge {
     }
 
     public boolean isBackup(List<Chat> backup) {
+        if (backup == null) {
+            return false;
+        }
         int indexmain = 0;
         int totchats = 0;
         for (Chat c : backup) {

--- a/iped-parsers/iped-parsers-impl/src/main/java/dpf/mg/udi/gpinf/whatsappextractor/WhatsAppParser.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/dpf/mg/udi/gpinf/whatsappextractor/WhatsAppParser.java
@@ -129,7 +129,7 @@ public class WhatsAppParser extends SQLite3DBParser {
 
     private static final boolean FALLBACK_FILENAME_APPROX_SIZE = true;
 
-    private static Pattern MSGSTORE_BKP = Pattern.compile("msgstore-\\d{4}-\\d{2}-\\d{2}"); //$NON-NLS-1$
+    private static final Pattern MSGSTORE_BKP = Pattern.compile("msgstore-\\d{4}-\\d{2}-\\d{2}"); //$NON-NLS-1$
     private static final String MSGSTORE_CRYPTO = "msgstore.db.crypt"; //$NON-NLS-1$
     private static final String IS_BACKUP_FROM = "isBackupFrom";
 
@@ -407,12 +407,25 @@ public class WhatsAppParser extends SQLite3DBParser {
         // list and parsing results won't be on memory yet
         findOtherDBS(searcher);
 
+        WhatsAppContext wcontext = dbsFound.get(DB.getId());
+        if (wcontext != null && wcontext.getChalist() == null) {
+            // if not parsed yet, parse the DB here
+            parseDB(wcontext, metadata, context, extFactory);
+        }
+
         // parse DBs found above
-        for (WhatsAppContext wcontext : dbsFound.values().toArray(new WhatsAppContext[0])) {
-            synchronized (wcontext) {
-                if (wcontext.getChalist() == null) {
+        for (WhatsAppContext other : dbsFound.values().toArray(new WhatsAppContext[0])) {
+            if (other == wcontext)
+                continue;
+            synchronized (other) {
+                if (other.getChalist() == null) {
                     // if not parsed yet, parse the DB here
-                    parseDB(wcontext, metadata, context, extFactory);
+                    try {
+                        parseDB(other, metadata, context, extFactory);
+                    } catch (Exception e) {
+                        logger.warn("could not parse DB {}", other.getItem().getName());
+                        // TODO: handle exception
+                    }
                 }
             }
         }
@@ -429,7 +442,6 @@ public class WhatsAppParser extends SQLite3DBParser {
             }
         });
 
-        WhatsAppContext wcontext = dbsFound.get(DB.getId());
         
         if (wcontext == null) {
             // MakePreviewTask enters here for main dbs and backups without main db

--- a/iped-parsers/iped-parsers-impl/src/main/java/dpf/mg/udi/gpinf/whatsappextractor/WhatsAppParser.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/dpf/mg/udi/gpinf/whatsappextractor/WhatsAppParser.java
@@ -410,7 +410,11 @@ public class WhatsAppParser extends SQLite3DBParser {
         WhatsAppContext wcontext = dbsFound.get(DB.getId());
         if (wcontext != null && wcontext.getChalist() == null) {
             // if not parsed yet, parse the DB here
-            parseDB(wcontext, metadata, context, extFactory);
+            synchronized (wcontext) {
+                if (wcontext.getChalist() == null) {
+                    parseDB(wcontext, metadata, context, extFactory);
+                }
+            }
         }
 
         // parse DBs found above

--- a/iped-parsers/iped-parsers-impl/src/main/java/dpf/mg/udi/gpinf/whatsappextractor/WhatsAppParser.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/dpf/mg/udi/gpinf/whatsappextractor/WhatsAppParser.java
@@ -427,8 +427,9 @@ public class WhatsAppParser extends SQLite3DBParser {
                     try {
                         parseDB(other, metadata, context, extFactory);
                     } catch (Exception e) {
-                        logger.warn("could not parse DB {}", other.getItem().getName());
-                        // TODO: handle exception
+                        logger.warn("Could not parse DB {} ({} bytes): {}", other.getItem().getPath(),
+                                other.getItem().getLength(), e.toString());
+                        logger.debug("", e);
                     }
                 }
             }

